src/ucs/arch/x86_64/cpu.c

```diff
@@ -588,6 +588,28 @@ static size_t ucs_cpu_memcpy_thresh(size_t user_val, size_t auto_val)
 }
 #endif
 
+static size_t ucs_cpu_nt_bt_thresh_min(size_t user_val)
+{
+    if (user_val != UCS_MEMUNITS_AUTO) {
+        return user_val;
+    }
+
+    if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
+        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 2) * 3);
+    } else {
+        return UCS_MEMUNITS_INF;
+    }
+}
+
+static size_t ucs_cpu_nt_dest_thresh()
+{
+    if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
+        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 5) * 29);
```

Comment:
why src thresh different from dst thresh?

Reply:
There is no "src thresh", only "dst_thresh". The variable "nt_buffer_transfer_min" is used to decide whether to use nt-buffer-transfer for a particular MPI job, not to control the selection of "ucs_x86_nt_src_buffer_transfer". It is set to "3/4" of the L3 cache size by default **mainly** to prevent the use of "nt-buffer-transfer" for Copy-in-Copy-out methods (eager-buffered) for workloads with ranks in the same L3. Users can override this by using "-x UCX_NT_BUFFER_TRANSFER_MIN=0" in the command line if their workload doesn't have ranks in the same L3.

The "dst_thresh" comes into play when the copy function "ucs_memcpy_relaxed" is called with a total_len close to the size of the L3 cache. In such cases, the hint "UCS_ARCH_MEMCPY_NT_DEST" is added based on the "dst_thresh" to avoid loading the destination data to the caches (L1, L2 and L3) and prevent cache pollution. This is because the total_len is close to the L3 cache size. The "dst_thresh" is currently set close to 90% of the L3 cache size.

Addition of hint happens here:
----------------------------------
/* Give precedence to non temporal destination */  
if (ucs_unlikely(total_len > ucs_global_opts.arch.nt_dest_threshold)) {  
    hint |= UCS_ARCH_MEMCPY_NT_DEST;  
}  

Setting of "dest_thresh":
--------------------------
static size_t ucs_cpu_nt_dest_thresh()
{
if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 5) * 29);
} else {
return UCS_MEMUNITS_INF;
}
}

When I share the osu_latency data for 4 different cases (ranks in same l3, ranks in different l3, ranks in different numa, ranks in different socket), I hope the selection criteria will become more clear.

Reply:
thanks, so dst thresh is used to dynamically hint when nt_buffer_transfer_min has been reached.